### PR TITLE
adds link to diffs page in footer of hex.outdated

### LINKED ISSUE #778: Allow to generate outdate list from project and upload it to hexdiff
Hey, I just saw Johanna presentation about hexdiff, amazing job with that!

I was wondering if it could be improved a little bit.

The use case would be this one: You have your project, and you run mix hex.outdated, now you have a list of packages that are outdated and needs an update. To be safe, you use hexdiff to verify what changed from each package. But doing so is tiresome as you need to get package by package and select the correct version in hexdiff interface manually.

My suggestion would be to add an option to mix hex.outdated that will generate a file with all the outdated packages, its current version, and the updated version.

And also an option in hexdiff website so we can upload that file, so hexdiff can automatically show the diff for all the packages in order, that way the user can save time.

Would something like that be possible/desirable?

Best regards,
Eduardo

### PR DESCRIPTION
**Pull request by supersimple**
Closes #778 

When there are outdated deps that can be upgraded, build a list of the current version and the latest version. Then use the hexpm API to create a short link.

diff

### CODE REVIEW

File: lib/mix/tasks/hex.outdated.ex [hunk 1]
```diff
@@ -247,4 +260,20 @@ defmodule Mix.Tasks.Hex.Outdated do
       Hex.Version.compare(lock, latest) == :lt
     end)
   end
+
+  defp maybe_diff_message([]), do: ""
+
+  defp maybe_diff_message(diff_links) do
+    "\nTo view the diffs in each available update, visit:\n" <>
```
Comment (ericmj) [inline lib/mix/tasks/hex.outdated.ex:267]:
```suggestion
    "\n\nTo view the diffs in each available update, visit:\n" <>
```

File: lib/mix/tasks/hex.outdated.ex [hunk 2]
```diff
@@ -153,19 +153,22 @@ defmodule Mix.Tasks.Hex.Outdated do
...
-      message =
+      base_message =
         "Up-to-date indicates if you have the latest version of a given package.\n" <>
           "Update possible indicates if your current requirement matches the latest version.\n" <>
           "Run `mix hex.outdated APP` to see requirements for a specific dependency."
 
-      Hex.Shell.info(["\n" | message])
+      diff_message = maybe_diff_message(diff_links)
+
+      Hex.Shell.info(["\n" | base_message <> diff_message])
```
Comment (ericmj) [inline lib/mix/tasks/hex.outdated.ex:171]:
```suggestion
      Hex.Shell.info(["\n", base_message, diff_message])
```

File: lib/mix/tasks/hex.outdated.ex [hunk 3]
```diff
@@ -239,6 +242,16 @@ defmodule Mix.Tasks.Hex.Outdated do
     ]
   end
 
+  defp build_diff_link([package, lock, latest, requirements]) do
+    outdated? = Hex.Version.compare(lock, latest) == :lt
+    req_matches? = Enum.all?(requirements, &version_match?(latest, &1))
+
+    case {outdated?, req_matches?} do
+      {true, true} -> "diffs[]=#{package}:#{lock}:#{latest}"
```
Comment (ericmj) [inline lib/mix/tasks/hex.outdated.ex:250]: Maybe we should show the diff even though the requirement doesn't match?

File: lib/hex/api/short_url.ex [hunk 4]
```diff
@@ -0,0 +1,12 @@
+defmodule Hex.API.ShortURL do
+  @moduledoc false
+
+  alias Hex.API
+
+  def create(url) do
+    case API.erlang_post_request(nil, "short_url", %{url: url}) do
+      {:ok, {201, %{"url" => short_url}, _resp_headers}} -> short_url
```
Comment (ericmj) [inline lib/hex/api/short_url.ex:8]:
```suggestion
      {:ok, {201, %{"url" => short_url}, _resp_headers}} -> {:ok, short_url}
```

File: lib/mix/tasks/hex.outdated.ex [hunk 5]
```diff
@@ -247,4 +260,20 @@ defmodule Mix.Tasks.Hex.Outdated do
...
+    "\nTo view the diffs in each available update, visit:\n" <>
+      diff_link(diff_links)
+  end
+
+  defp diff_link(diff_links) do
+    long_url = "https://diff.hex.pm/diffs?" <> Enum.join(diff_links, "&")
+
+    case Hex.API.ShortURL.create(long_url) do
+      :error -> long_url
+      short_url -> short_url
```
Comment (ericmj) [inline lib/mix/tasks/hex.outdated.ex:276]:
```suggestion
      {:ok, short_url} -> short_url
```

File: lib/mix/tasks/hex.outdated.ex (hunk 3 above)
Comment (supersimple) [reply lib/mix/tasks/hex.outdated.ex:250]: maybe. I worry about signal vs noise ratio if we do that. the ones we show here are actionable (presumably.) @wojtekmach @joladev WDYT?

**Comment by cw789** [comment]
Just found this new feature. It's sooo cool.
Thank you very very much. & thanks for everything else.

**Pull request merged**